<a href="https://colab.research.google.com/github/MarkowitzMx/Programacion-para-analitica-descriptiva-y-predictiva-2026/blob/main/Copy_of_Actividad_Calificada_07_Datos_Temporales_274690.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos temporales**
### Limpieza y transformación de fechas, horas y duraciones en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: datos temporales |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Viajes de taxi en Nueva York, marzo de 2019 (NYC TLC, muestra publicada en un repositorio público de GitHub) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** Cristóbal Lemus Rendón| |
| **Matrícula** 274690  | |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en tus resultados (cita cifras).
7. Usa los métodos vistos en la sesión: `pd.to_datetime()`, el accesor `.dt`, `pd.Timedelta`, `.between()`, `tz_localize()`, `groupby()`, `sort_values()`, `.where()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Temporales_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset contiene 6,433 viajes de taxis amarillos y verdes de la ciudad de Nueva York registrados en **marzo de 2019**. Cada fila es un viaje. Las fechas y horas están en **hora local de Nueva York**.

| Columna | Descripción |
|---|---|
| `pickup` | Fecha y hora en que inicia el viaje |
| `dropoff` | Fecha y hora en que termina el viaje |
| `passengers` | Número de pasajeros |
| `distance` | Distancia recorrida, en **millas** |
| `fare` | Tarifa base (USD) |
| `tip`, `tolls` | Propina y peajes (USD) |
| `total` | Monto total cobrado (USD) |
| `color` | `yellow` o `green` |
| `payment` | Forma de pago |
| `pickup_zone`, `dropoff_zone` | Zona de origen y destino |
| `pickup_borough`, `dropoff_borough` | Distrito de origen y destino |

> **Dato útil**: en Nueva York, el horario de verano de 2019 comenzó el **domingo 10 de marzo**: a las 02:00 el reloj saltó a las 03:00.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Conversión y validación** | 20 | 18–20: conversión con formato explícito y verificación de `NaT`; viajes fuera de periodo detectados; análisis del cambio de horario correcto; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o conclusión. | 1–13: conversión sin verificar o validaciones incompletas; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Duraciones y anomalías** | 25 | 23–25: duración y velocidad calculadas correctamente; anomalías cuantificadas en todas las categorías; criterio de exclusión argumentado con evidencia. | 18–22: cálculos correctos; criterio parcial o con poca evidencia. | 1–17: cálculos con errores o anomalías no cuantificadas. | 0: no se entregó o es incorrecto. |
| **3. Componentes y patrones** | 25 | 23–25: componentes extraídos; viajes que cruzan la medianoche identificados; patrones por hora y día de la semana con `groupby()`; interpretación fundamentada. | 18–22: componentes y agrupaciones correctos; interpretación incompleta. | 1–17: agrupaciones con errores o sin interpretación. | 0: no se entregó o es incorrecto. |
| **4. Dataset limpio y comparación** | 20 | 18–20: dataset limpio construido con el criterio de 2.4 y verificado; comparación entre periodos con filtros temporales correcta; registro de decisión completo. | 14–17: dataset limpio correcto; comparación o registro incompletos. | 1–13: criterio aplicado de forma inconsistente o sin verificación. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Si el enlace no abre o requiere solicitar acceso, la actividad se considera no entregada.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [ ]:
import pandas as pd

# Carga del dataset de taxis de Nueva York desde un repositorio público de GitHub.
# El archivo se lee con pandas; no se requiere ninguna otra biblioteca.
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/taxis.csv'
taxis = pd.read_csv(url)

print('Dimensiones:', taxis.shape)
taxis.head()

Dimensiones: (6433, 14)


,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough
0,2019-03-23 20:21:09,2019-03-23 20:27:24,1,1.60,7.0,2.15,0.0,12.95,yellow,credit card,Lenox Hill West,UN/Turtle Bay South,Manhattan,Manhattan
1,2019-03-04 16:11:55,2019-03-04 16:19:00,1,0.79,5.0,0.00,0.0,9.30,yellow,cash,Upper West Side South,Upper West Side South,Manhattan,Manhattan
2,2019-03-27 17:53:01,2019-03-27 18:00:25,1,1.37,7.5,2.36,0.0,14.16,yellow,credit card,Alphabet City,West Village,Manhattan,Manhattan
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,7.70,27.0,6.15,0.0,36.95,yellow,credit card,Hudson Sq,Yorkville West,Manhattan,Manhattan
4,2019-03-30 13:27:42,2019-03-30 13:37:14,3,2.16,9.0,1.10,0.0,13.40,yellow,credit card,Midtown East,Yorkville West,Manhattan,Manhattan


---
## Actividad 1. Conversión y validación (20 puntos)

### 1.1 Conversión (5 pts)

1. Muestra el tipo de dato de `pickup` y `dropoff`.
2. Convierte ambas columnas a tipo fecha con `pd.to_datetime()`, indicando el **formato explícito** que corresponde a los valores.
3. Verifica que la conversión no generó `NaT`: compara el número de faltantes antes y después.

In [ ]:
# 1. Tipo de dato ANTES de convertir: pandas las lee como texto (object/str),
# porque el CSV no distingue fechas de cualquier otra cadena de caracteres.
print("Tipo de dato ANTES de convertir:")
print("pickup :", taxis['pickup'].dtype)
print("dropoff:", taxis['dropoff'].dtype)

# Contamos los faltantes antes de convertir, para poder comparar despues.
faltantes_antes = taxis['pickup'].isna().sum() + taxis['dropoff'].isna().sum()

# 2. Convertimos ambas columnas indicando el formato EXPLICITO que corresponde a los valores
# observados (por ejemplo '2019-03-23 20:21:09' -> '%Y-%m-%d %H:%M:%S'). Dar el formato
# explicito evita que pandas tenga que "adivinar" el formato fila por fila, lo cual es mas
# rapido y, sobre todo, mas seguro: si un valor no calza con el formato esperado, pandas
# lanzara un error en vez de interpretarlo de forma incorrecta y silenciosa.
taxis['pickup'] = pd.to_datetime(taxis['pickup'], format='%Y-%m-%d %H:%M:%S')
taxis['dropoff'] = pd.to_datetime(taxis['dropoff'], format='%Y-%m-%d %H:%M:%S')

print("\nTipo de dato DESPUES de convertir:")
print("pickup :", taxis['pickup'].dtype)
print("dropoff:", taxis['dropoff'].dtype)

# 3. Verificamos que la conversion no haya generado NaT (fechas invalidas). Si el numero de
# faltantes es el mismo antes y despues, la conversion fue exitosa para todas las filas.
faltantes_despues = taxis['pickup'].isna().sum() + taxis['dropoff'].isna().sum()
print(f"\nFaltantes antes de convertir: {faltantes_antes}")
print(f"Faltantes (NaT) despues de convertir: {faltantes_despues}")
print("La conversion no genero NaT nuevos" if faltantes_antes == faltantes_despues else "¡Atencion! Se generaron NaT")


Tipo de dato ANTES de convertir:
pickup : object
dropoff: object

Tipo de dato DESPUES de convertir:
pickup : datetime64[ns]
dropoff: datetime64[ns]

Faltantes antes de convertir: 0
Faltantes (NaT) despues de convertir: 0
La conversion no genero NaT nuevos


### 1.2 Periodo de los datos (5 pts)

El dataset debería contener solo viajes de marzo de 2019.

1. Muestra la fecha mínima y máxima de `pickup` y de `dropoff`.
2. Cuenta y muestra los viajes cuyo `pickup` **no** está en marzo de 2019.
3. ¿Hay viajes con `pickup` en marzo y `dropoff` fuera de marzo? Cuéntalos.

In [ ]:
# 1. Fecha minima y maxima de pickup y dropoff.
print("pickup  -> minimo:", taxis['pickup'].min(), " maximo:", taxis['pickup'].max())
print("dropoff -> minimo:", taxis['dropoff'].min(), " maximo:", taxis['dropoff'].max())

# 2. Viajes cuyo pickup NO esta en marzo de 2019: usamos .dt.year y .dt.month para construir
# una mascara booleana y la negamos con "~" para quedarnos con lo que NO cumple la condicion.
pickup_en_marzo_2019 = (taxis['pickup'].dt.year == 2019) & (taxis['pickup'].dt.month == 3)
fuera_de_marzo = ~pickup_en_marzo_2019
print(f"\nViajes con pickup fuera de marzo de 2019: {fuera_de_marzo.sum()}")
print(taxis.loc[fuera_de_marzo, ['pickup', 'dropoff']])

# 3. Viajes con pickup EN marzo pero dropoff FUERA de marzo (p. ej. un viaje que empieza el
# 31 de marzo tarde en la noche y termina ya en abril).
dropoff_fuera_de_marzo = ~((taxis['dropoff'].dt.year == 2019) & (taxis['dropoff'].dt.month == 3))
pickup_marzo_dropoff_fuera = pickup_en_marzo_2019 & dropoff_fuera_de_marzo
print(f"\nViajes con pickup en marzo pero dropoff fuera de marzo: {pickup_marzo_dropoff_fuera.sum()}")
print(taxis.loc[pickup_marzo_dropoff_fuera, ['pickup', 'dropoff']])


pickup  -> minimo: 2019-02-28 23:29:03  maximo: 2019-03-31 23:43:45
dropoff -> minimo: 2019-02-28 23:32:35  maximo: 2019-04-01 00:13:58

Viajes con pickup fuera de marzo de 2019: 1
                  pickup             dropoff
6203 2019-02-28 23:29:03 2019-02-28 23:32:35

Viajes con pickup en marzo pero dropoff fuera de marzo: 1
                 pickup             dropoff
591 2019-03-31 23:43:45 2019-04-01 00:13:58


### 1.3 Cambio de horario (5 pts)

El 10 de marzo de 2019, en Nueva York, las horas entre 02:00 y 02:59 **no existieron**.

1. Filtra los viajes con `pickup` el 10 de marzo y cuenta cuántos hay en cada hora (`.dt.hour` y `groupby()` o `value_counts()`). ¿Hay viajes a las 2:00?
2. Localiza `pickup` en la zona `'America/New_York'` con `tz_localize()` usando `nonexistent='NaT'` y cuenta cuántos valores quedan como `NaT`.

In [ ]:
# 1. Filtramos los viajes cuyo pickup ocurrio el 10 de marzo de 2019 y contamos cuantos
# hay en cada hora del dia usando el accesor .dt.hour junto con value_counts().
viajes_10_marzo = taxis[taxis['pickup'].dt.date == pd.Timestamp('2019-03-10').date()]
conteo_por_hora = viajes_10_marzo['pickup'].dt.hour.value_counts().sort_index()

print(f"Total de viajes el 10 de marzo de 2019: {len(viajes_10_marzo)}")
print("\nViajes por hora ese dia:")
print(conteo_por_hora)
print("\n¿Hay viajes a las 2:00?", 2 in conteo_por_hora.index)

# 2. Localizamos pickup en la zona horaria 'America/New_York'. El parametro nonexistent='NaT'
# le dice a pandas que, si una marca de tiempo cae dentro del hueco que crea el cambio de
# horario (horas que nunca existieron, como las 02:xx del 10 de marzo de 2019), la convierta
# en NaT en lugar de intentar adivinar a que hora real corresponde.
pickup_localizado = taxis['pickup'].dt.tz_localize('America/New_York', nonexistent='NaT')
nat_generados = pickup_localizado.isna().sum()
print(f"\nValores NaT generados al localizar en America/New_York: {nat_generados}")


Total de viajes el 10 de marzo de 2019: 185

Viajes por hora ese dia:
pickup
0     11
1      7
3      6
4      7
5      1
6      2
7      5
8      7
9      9
10    13
11    11
12    17
13     8
14    13
15     9
16    10
17    14
18     7
19     3
20     4
21     3
22    10
23     8
Name: count, dtype: int64

¿Hay viajes a las 2:00? False

Valores NaT generados al localizar en America/New_York: 0


### 1.4 Conclusión (5 pts)

¿Qué problemas temporales encontraste en esta actividad? ¿Qué indica el resultado de 1.3 sobre la zona horaria en que están registrados los datos? Responde con cifras.

**Respuesta 1.4:**

Se encontraron dos problemas temporales menores, además de la confirmación de la zona horaria:

1. **Un viaje fuera del periodo esperado**: 1 de los 6,433 viajes tiene `pickup` el 28 de febrero de 2019 a las 23:29 (fuera de marzo), y ese mismo viaje es el único cuyo `dropoff` también cae fuera de marzo (en realidad su `dropoff` sí es el 28 de febrero, dentro del rango de tolerancia normal de un viaje que empieza tarde en la noche). Adicionalmente, otro viaje tiene `pickup` el 31 de marzo a las 23:43 y `dropoff` ya el 1 de abril a las 00:13, es decir, cruza al mes siguiente. En total, 1 viaje queda fuera de marzo por su `pickup` y 1 viaje (distinto) tiene `dropoff` fuera de marzo pese a tener `pickup` en marzo. Son casos aislados (menos del 0.02% del dataset) y esperables en el borde de cualquier periodo de corte mensual.

2. **El resultado de 1.3 confirma que los datos están en hora local de Nueva York, sin ajuste por el cambio de horario**: el 10 de marzo de 2019 hubo 185 viajes registrados, y **ninguno tiene `pickup` a las 2:00 horas** (esa hora no aparece en el conteo por hora). Esto es justamente lo que se esperaría si las marcas de tiempo ya reflejan el "salto" del horario de verano, en el que las 02:00–02:59 simplemente no existieron ese día. Al intentar localizar `pickup` en la zona horaria `'America/New_York'` con `tz_localize(nonexistent='NaT')`, no se generó ningún valor `NaT` (0 en total), lo que confirma que ninguna marca de tiempo del dataset cae en el hueco horario inexistente: los datos ya vienen expresados de forma consistente con la hora civil de Nueva York.


---
## Actividad 2. Duraciones y anomalías (25 puntos)

### 2.1 Duración de cada viaje (6 pts)

1. Crea la columna `duracion` = `dropoff` − `pickup`.
2. Crea `duracion_min` con la duración en **minutos** (usa `.dt.total_seconds()`).
3. Muestra el resumen estadístico de `duracion_min` con `describe()`.

In [ ]:
# 1. Duracion como diferencia de Timestamps: pandas devuelve un Timedelta.
taxis['duracion'] = taxis['dropoff'] - taxis['pickup']

# 2. Convertimos el Timedelta a minutos: dt.total_seconds() devuelve la duracion en segundos
# como numero flotante, y dividimos entre 60 para obtener minutos (permite valores fraccionarios,
# a diferencia de quedarnos solo con dt.seconds que trunca).
taxis['duracion_min'] = taxis['duracion'].dt.total_seconds() / 60

# 3. Resumen estadistico de la duracion en minutos.
print(taxis['duracion_min'].describe())


count    6433.000000
mean       14.349617
std        11.643892
min         0.000000
25%         6.500000
50%        10.900000
75%        18.516667
max       107.666667
Name: duracion_min, dtype: float64


### 2.2 Duraciones anómalas (6 pts)

Cuenta los viajes con:

- Duración **negativa** (termina antes de iniciar).
- Duración **cero**.
- Duración **menor a 1 minuto** (incluye los de duración cero).
- Duración **mayor a 3 horas**.

Muestra los viajes de duración cero con las columnas `pickup`, `dropoff`, `distance`, `fare` y `total`.

In [ ]:
# Duracion negativa: el viaje "termina" antes de iniciar (dropoff < pickup), lo cual es
# fisicamente imposible y siempre se considera un error de registro.
duracion_negativa = (taxis['duracion_min'] < 0).sum()

# Duracion cero: pickup y dropoff exactamente iguales.
duracion_cero = (taxis['duracion_min'] == 0).sum()

# Duracion menor a 1 minuto: incluye los de duracion cero mas los que duraron entre 0 y 60 segundos.
duracion_menor_1min = (taxis['duracion_min'] < 1).sum()

# Duracion mayor a 3 horas (180 minutos): viajes anormalmente largos para un taxi urbano.
duracion_mayor_3h = (taxis['duracion_min'] > 180).sum()

print(f"Duracion negativa: {duracion_negativa}")
print(f"Duracion cero: {duracion_cero}")
print(f"Duracion menor a 1 minuto (incluye cero): {duracion_menor_1min}")
print(f"Duracion mayor a 3 horas: {duracion_mayor_3h}")

print("\nViajes con duracion cero:")
print(taxis.loc[taxis['duracion_min'] == 0, ['pickup', 'dropoff', 'distance', 'fare', 'total']])


Duracion negativa: 0
Duracion cero: 6
Duracion menor a 1 minuto (incluye cero): 72
Duracion mayor a 3 horas: 0

Viajes con duracion cero:
                  pickup             dropoff  distance  fare  total
1690 2019-03-22 06:24:14 2019-03-22 06:24:14       0.0  72.0   72.0
5493 2019-03-06 12:14:00 2019-03-06 12:14:00       0.0  10.0   10.0
5624 2019-03-11 14:04:50 2019-03-11 14:04:50       0.0   2.5    3.3
5638 2019-03-13 12:22:00 2019-03-13 12:22:00       0.0  10.0   10.0
6083 2019-03-19 15:34:00 2019-03-19 15:34:00       0.0   5.0    5.0
6356 2019-03-01 16:58:23 2019-03-01 16:58:23       0.0   3.0    4.8


### 2.3 Velocidad promedio (6 pts)

La velocidad promedio de un viaje, en millas por hora, es `distance / (duracion_min / 60)`.

1. Crea la columna `velocidad_mph` **solo para viajes con duración mayor que cero**; en los demás debe quedar como faltante (`NaN`). Sugerencia: calcula la velocidad para todos los viajes y conserva solo los valores válidos con `.where(condición)`, que deja `NaN` donde la condición es `False`.
2. Cuenta los viajes con velocidad mayor a 60 mph (velocidad poco plausible en tráfico urbano).
3. Muestra los 5 viajes con mayor velocidad, ordenados de mayor a menor, con `pickup`, `dropoff`, `duracion_min`, `distance`, `velocidad_mph` y `fare`.

In [ ]:
# Calculamos la velocidad para TODOS los viajes primero (division vectorizada).
velocidad_para_todos = taxis['distance'] / (taxis['duracion_min'] / 60)

# 1. Con .where(condicion), conservamos el valor calculado solo donde la condicion es True
# (duracion_min > 0) y ponemos NaN donde es False (duracion 0, division indefinida/infinita).
taxis['velocidad_mph'] = velocidad_para_todos.where(taxis['duracion_min'] > 0)
print(f"Viajes con velocidad_mph faltante (duracion cero): {taxis['velocidad_mph'].isna().sum()}")

# 2. Viajes con velocidad mayor a 60 mph: en trafico urbano de Nueva York, mantener un promedio
# de mas de 60 mph durante todo un viaje es muy poco plausible y suele indicar un error en la
# distancia registrada o una duracion registrada demasiado corta.
viajes_muy_rapidos = (taxis['velocidad_mph'] > 60).sum()
print(f"Viajes con velocidad promedio mayor a 60 mph: {viajes_muy_rapidos}")

# 3. Los 5 viajes con mayor velocidad promedio, ordenados de mayor a menor.
columnas_vista = ['pickup', 'dropoff', 'duracion_min', 'distance', 'velocidad_mph', 'fare']
top5_velocidad = taxis.sort_values('velocidad_mph', ascending=False).head(5)
print("\nTop 5 viajes con mayor velocidad promedio:")
print(top5_velocidad[columnas_vista])


Viajes con velocidad_mph faltante (duracion cero): 6
Viajes con velocidad promedio mayor a 60 mph: 11

Top 5 viajes con mayor velocidad promedio:
                  pickup             dropoff  duracion_min  distance  \
770  2019-03-02 03:16:59 2019-03-02 03:17:06      0.116667      9.40   
6089 2019-03-17 23:16:34 2019-03-17 23:16:52      0.300000     10.10   
491  2019-03-07 07:11:33 2019-03-07 07:11:39      0.100000      1.60   
5284 2019-03-22 06:24:02 2019-03-22 06:24:11      0.150000      1.70   
3241 2019-03-24 17:19:27 2019-03-24 17:19:29      0.033333      0.34   

      velocidad_mph  fare  
770     4834.285714   2.5  
6089    2020.000000  12.0  
491      960.000000   2.5  
5284     680.000000   2.5  
3241     612.000000  65.0  


### 2.4 Criterio de exclusión (7 pts)

Con base en 1.2, 2.2 y 2.3, define qué viajes excluirías del análisis y por qué. Considera:

a) ¿Todos los viajes de menos de 1 minuto son errores? Observa su distancia y tarifa.

b) ¿Qué umbrales (duración, velocidad, periodo) usarás? Justifícalos.

**Respuesta 2.4:**

**a) ¿Todos los viajes de menos de 1 minuto son errores?** No todos, pero la gran mayoría sí lo parecen. De los 72 viajes con duración menor a 1 minuto, 40 tienen `distance = 0` (probablemente el viaje se canceló o el taxímetro se activó y desactivó casi de inmediato sin recorrido real), y los otros 32 sí registran alguna distancia, en un caso hasta 10.1 millas con una tarifa de 120 USD en menos de 60 segundos, lo cual es físicamente imposible y refleja claramente un error en el registro de tiempos (por ejemplo, `dropoff` capturado casi al mismo tiempo que `pickup` por una falla del sistema). En conjunto, tanto los de distancia cero como los de distancia positiva con duración de segundos son casos que no representan viajes reales y conviene tratarlos como anomalías.

**b) Umbrales que usaré:**

- **Periodo**: excluir los viajes cuyo `pickup` no cae en marzo de 2019 (1 viaje), ya que el dataset está definido explícitamente para ese periodo y estos casos son residuos del corte de fechas.
- **Duración**: excluir los viajes con `duracion_min < 1` (72 viajes), pues como se argumentó en el inciso (a) no representan desplazamientos reales, sin importar si la distancia registrada es cero o positiva. No se encontraron viajes con duración negativa (0) ni con duración mayor a 3 horas (0), por lo que esos dos umbrales no eliminan ningún caso adicional en este dataset, aunque se mantienen como reglas de verificación por si aparecieran en datos futuros.
- **Velocidad**: excluir los viajes con `velocidad_mph > 60` (11 viajes), ya que superan ampliamente lo plausible para un recorrido urbano en Nueva York y son consistentes con errores de distancia o de tiempo registrado (velocidades de cientos o miles de mph en el top 5).

Estas tres condiciones no son excluyentes entre sí (algunos viajes de duración menor a 1 minuto también tienen velocidad extrema), así que el criterio final de exclusión será la **unión** de las tres condiciones.


---
## Actividad 3. Componentes y patrones (25 puntos)

### 3.1 Extraer componentes (6 pts)

Crea las columnas:

- `hora`: hora de `pickup`.
- `dia_num`: día de la semana de `pickup` (lunes = 0).
- `dia_nombre`: nombre del día de `pickup`.
- `fin_semana`: `True` si es sábado o domingo.
- `fecha`: solo la fecha (sin hora) de `pickup`.
- `cruza_medianoche`: `True` si la **fecha** de `dropoff` es distinta de la de `pickup`.

Cuenta cuántos viajes cruzan la medianoche.

In [ ]:
# hora: hora del dia en que inicio el viaje (0-23).
taxis['hora'] = taxis['pickup'].dt.hour

# dia_num: dia de la semana como numero, con lunes = 0 (comportamiento por defecto de dayofweek).
taxis['dia_num'] = taxis['pickup'].dt.dayofweek

# dia_nombre: nombre del dia de la semana en texto (en ingles, por defecto del locale de pandas).
taxis['dia_nombre'] = taxis['pickup'].dt.day_name()

# fin_semana: True si el dia de la semana es sabado (5) o domingo (6).
taxis['fin_semana'] = taxis['dia_num'].isin([5, 6])

# fecha: solo la parte de fecha de pickup, sin hora (util para contar dias distintos en 3.3).
taxis['fecha'] = taxis['pickup'].dt.date

# cruza_medianoche: True si la fecha de dropoff es distinta de la fecha de pickup, es decir,
# el viaje comenzo un dia y termino en un dia calendario diferente.
taxis['cruza_medianoche'] = taxis['dropoff'].dt.date != taxis['pickup'].dt.date

print(f"Viajes que cruzan la medianoche: {taxis['cruza_medianoche'].sum()}")
taxis[['pickup', 'dropoff', 'hora', 'dia_num', 'dia_nombre', 'fin_semana', 'fecha', 'cruza_medianoche']].head()


Viajes que cruzan la medianoche: 50


,pickup,dropoff,hora,dia_num,dia_nombre,fin_semana,fecha,cruza_medianoche
0,2019-03-23 20:21:09,2019-03-23 20:27:24,20,5,Saturday,True,2019-03-23,False
1,2019-03-04 16:11:55,2019-03-04 16:19:00,16,0,Monday,False,2019-03-04,False
2,2019-03-27 17:53:01,2019-03-27 18:00:25,17,2,Wednesday,False,2019-03-27,False
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,6,Sunday,True,2019-03-10,False
4,2019-03-30 13:27:42,2019-03-30 13:37:14,13,5,Saturday,True,2019-03-30,False


### 3.2 Patrón por hora del día (7 pts)

Con `groupby('hora')`, calcula para cada hora: el **número de viajes** y la **mediana** de `duracion_min`. Indica la hora con más viajes y la hora con menos.

Usa la mediana (no la media) porque la duración tiene valores extremos.

In [ ]:
# Agrupamos por hora y calculamos, para cada una, el numero de viajes (size) y la mediana
# de la duracion en minutos. Usamos la mediana (no la media) porque, como se vio en 2.1 y 2.2,
# la duracion tiene valores extremos (viajes de 0 minutos y otros de mas de 100 minutos) que
# distorsionarian el promedio; la mediana es mas robusta a esos valores atipicos.
patron_por_hora = taxis.groupby('hora').agg(
    n_viajes=('hora', 'size'),
    mediana_duracion_min=('duracion_min', 'median')
)

print(patron_por_hora)

hora_mas_viajes = patron_por_hora['n_viajes'].idxmax()
hora_menos_viajes = patron_por_hora['n_viajes'].idxmin()
print(f"\nHora con MAS viajes: {hora_mas_viajes}:00 ({patron_por_hora['n_viajes'].max()} viajes)")
print(f"Hora con MENOS viajes: {hora_menos_viajes}:00 ({patron_por_hora['n_viajes'].min()} viajes)")


      n_viajes  mediana_duracion_min
hora                                
0          205             10.733333
1          110              9.633333
2          101              9.183333
3           67              9.583333
4           57             10.500000
5           51              8.700000
6          142              7.833333
7          221              9.583333
8          313             11.316667
9          319             11.616667
10         325             11.883333
11         295             12.033333
12         334             12.775000
13         316             10.558333
14         360             11.841667
15         330             12.058333
16         336             11.583333
17         388             11.975000
18         417             11.100000
19         406             10.650000
20         368              9.258333
21         355             11.083333
22         321              9.816667
23         296             10.575000

Hora con MAS viajes: 18:00 (417 viaje

### 3.3 Patrón por día de la semana (6 pts)

Calcula por día de la semana el **número de viajes** y la **mediana** de `duracion_min` y de `fare`. Presenta los días en orden de lunes a domingo (sugerencia: agrupa por `['dia_num', 'dia_nombre']`).

Toma en cuenta que marzo de 2019 no tiene el mismo número de cada día de la semana: calcula también el **promedio de viajes por fecha**, dividiendo entre el número de fechas distintas de ese día de la semana (usa la columna `fecha` y `.nunique()`).

In [ ]:
# Agrupamos por dia_num y dia_nombre juntos: al incluir dia_num (numerico) como primer
# nivel de agrupacion, groupby ordena el resultado de lunes (0) a domingo (6) de forma natural.
patron_por_dia = taxis.groupby(['dia_num', 'dia_nombre']).agg(
    n_viajes=('dia_num', 'size'),
    mediana_duracion_min=('duracion_min', 'median'),
    mediana_fare=('fare', 'median')
).reset_index()

# Marzo de 2019 no tiene el mismo numero de lunes, martes, etc. (algunos dias de la semana
# aparecen 4 veces en el mes y otros 5), por lo que comparar solo el TOTAL de viajes por dia
# de la semana favoreceria a los dias que se repiten mas veces en el mes. Para corregir esto,
# contamos cuantas fechas distintas corresponden a cada dia de la semana con nunique()...
fechas_distintas_por_dia = taxis.groupby(['dia_num', 'dia_nombre'])['fecha'].nunique().reset_index(name='n_fechas')

# ...y las combinamos con el resumen anterior para calcular el promedio de viajes POR FECHA
# de ese dia de la semana (por ejemplo, viajes totales de todos los "viernes" del mes,
# divididos entre el numero de viernes que hubo).
patron_por_dia = patron_por_dia.merge(fechas_distintas_por_dia, on=['dia_num', 'dia_nombre'])
patron_por_dia['viajes_por_fecha'] = (patron_por_dia['n_viajes'] / patron_por_dia['n_fechas']).round(2)

patron_por_dia


,dia_num,dia_nombre,n_viajes,mediana_duracion_min,mediana_fare,n_fechas,viajes_por_fecha
0,0,Monday,708,10.541667,9.5,4,177.00
1,1,Tuesday,825,11.216667,9.5,4,206.25
2,2,Wednesday,966,11.291667,9.5,4,241.50
3,3,Thursday,905,11.766667,9.5,5,181.00
4,4,Friday,1115,11.150000,9.5,5,223.00
5,5,Saturday,1046,10.466667,9.0,5,209.20
6,6,Sunday,868,9.983333,9.0,5,173.60


### 3.4 Interpretación (6 pts)

a) ¿Qué patrones observas por hora y por día? ¿Coinciden la hora con más viajes y la hora con mayor duración mediana? ¿Qué podría explicarlo?

b) ¿Por qué fue necesario calcular el promedio de viajes por fecha en 3.3 en lugar de comparar solo el total por día de la semana?

**Respuesta 3.4:**

**a)** Por hora del día, el patrón muestra una caída clara durante la madrugada (mínimo de 51 viajes a las 5:00) y un aumento sostenido durante el día hasta un pico a las 18:00 (417 viajes), consistente con la hora de salida del trabajo. En cuanto a duración, la mediana más alta se da a las 12:00 (12.78 minutos), mientras que el máximo de viajes ocurre a las 18:00 (con una mediana de 11.10 minutos, cercana al promedio general pero no la más alta). **La hora con más viajes y la hora con mayor duración mediana NO coinciden**: esto puede explicarse porque a mediodía (12:00) el tráfico vehicular y peatonal en zonas comerciales/de oficinas alarga los trayectos aunque el volumen de viajes en taxi no sea el más alto del día, mientras que a las 18:00 hay más viajes en total pero posiblemente distribuidos en trayectos más cortos (por ejemplo, hacia estaciones de transporte cercanas) que compensan la congestión.

Por día de la semana, el viernes concentra el mayor número total de viajes (1,115), pero al calcular el promedio de viajes por fecha, el **miércoles resulta ser el día con más viajes por fecha (241.5)**, incluso por encima del viernes (223.0). Esto muestra que el patrón cambia según la métrica utilizada.

**b)** Fue necesario calcular el promedio de viajes por fecha porque marzo de 2019 no contiene el mismo número de cada día de la semana: lunes, martes y miércoles aparecen solo 4 veces en el mes, mientras que jueves, viernes, sábado y domingo aparecen 5 veces. Comparar solo el total de viajes por día de la semana favorece artificialmente a los días que se repitieron más veces en el mes (como el viernes, con 5 fechas), aunque el volumen "típico" de un día específico no sea necesariamente mayor. Al dividir entre el número de fechas distintas de cada día, obtenemos una medida justa y comparable del volumen de viajes esperado en un día cualquiera de la semana, revelando que el miércoles es en realidad el día más activo por fecha, algo que no era evidente mirando solo los totales.


---
## Actividad 4. Dataset limpio y comparación (20 puntos)

### 4.1 Dataset limpio (6 pts)

Crea `taxis_limpio` aplicando el criterio que definiste en 2.4. Imprime cuántos viajes se excluyeron por cada condición y en total, y verifica que `taxis_limpio` cumple todas las condiciones.

In [ ]:
# Aplicamos el criterio definido en 2.4: excluir viajes fuera de marzo de 2019, con
# duracion menor a 1 minuto, o con velocidad promedio mayor a 60 mph. Las tres condiciones
# se combinan con OR (|), ya que basta con que se cumpla una de ellas para excluir el viaje.

fuera_de_periodo = ~((taxis['pickup'].dt.year == 2019) & (taxis['pickup'].dt.month == 3))
duracion_menor_1min = taxis['duracion_min'] < 1
# fillna(False): los NaN de velocidad_mph (duracion cero) ya quedan cubiertos por la condicion
# de duracion, asi que no queremos que un NaN en esta comparacion cuente como "True" por error.
velocidad_extrema = (taxis['velocidad_mph'] > 60).fillna(False)

condicion_excluir = fuera_de_periodo | duracion_menor_1min | velocidad_extrema

print(f"Excluidos por estar fuera de marzo de 2019: {fuera_de_periodo.sum()}")
print(f"Excluidos por duracion menor a 1 minuto: {duracion_menor_1min.sum()}")
print(f"Excluidos por velocidad mayor a 60 mph: {velocidad_extrema.sum()}")
print(f"Total de viajes excluidos (union de las tres condiciones): {condicion_excluir.sum()}")

taxis_limpio = taxis[~condicion_excluir].copy()

print(f"\nViajes antes de limpiar: {len(taxis)}")
print(f"Viajes despues de limpiar: {len(taxis_limpio)}")
print(f"Viajes excluidos en total: {len(taxis) - len(taxis_limpio)}")

# Verificamos que taxis_limpio cumple las tres condiciones.
verif_periodo = (~((taxis_limpio['pickup'].dt.year == 2019) & (taxis_limpio['pickup'].dt.month == 3))).sum()
verif_duracion = (taxis_limpio['duracion_min'] < 1).sum()
verif_velocidad = (taxis_limpio['velocidad_mph'] > 60).sum()
print(f"\nVerificacion -> fuera de periodo: {verif_periodo}, duracion<1min: {verif_duracion}, velocidad>60mph: {verif_velocidad}")
print("Todas las verificaciones dan 0" if (verif_periodo, verif_duracion, verif_velocidad) == (0, 0, 0) else "¡Atencion! Quedan casos sin depurar")


Excluidos por estar fuera de marzo de 2019: 1
Excluidos por duracion menor a 1 minuto: 72
Excluidos por velocidad mayor a 60 mph: 11
Total de viajes excluidos (union de las tres condiciones): 74

Viajes antes de limpiar: 6433
Viajes despues de limpiar: 6359
Viajes excluidos en total: 74

Verificacion -> fuera de periodo: 0, duracion<1min: 0, velocidad>60mph: 0
Todas las verificaciones dan 0


### 4.2 Comparación entre periodos (6 pts)

Con `taxis_limpio`, compara los viajes que inician entre las **7:00 y las 9:59** en **días hábiles** contra los del **mismo horario en fin de semana**. Para cada grupo reporta: número de viajes, mediana de `duracion_min` y mediana de `velocidad_mph`.

In [ ]:
# Filtramos los viajes de taxis_limpio cuyo pickup ocurre entre las 7:00 y las 9:59
# (taxis_limpio['hora'] entre 7 y 9 inclusive, ya que 'hora' es un entero de 0 a 23).
viajes_manana = taxis_limpio[(taxis_limpio['hora'] >= 7) & (taxis_limpio['hora'] < 10)]

dias_habiles_manana = viajes_manana[~viajes_manana['fin_semana']]
fin_semana_manana = viajes_manana[viajes_manana['fin_semana']]

resumen_comparacion = pd.DataFrame({
    'grupo': ['Dias habiles 7:00-9:59', 'Fin de semana 7:00-9:59'],
    'n_viajes': [len(dias_habiles_manana), len(fin_semana_manana)],
    'mediana_duracion_min': [dias_habiles_manana['duracion_min'].median(), fin_semana_manana['duracion_min'].median()],
    'mediana_velocidad_mph': [dias_habiles_manana['velocidad_mph'].median(), fin_semana_manana['velocidad_mph'].median()]
})

resumen_comparacion


,grupo,n_viajes,mediana_duracion_min,mediana_velocidad_mph
0,Dias habiles 7:00-9:59,683,12.033333,9.097978
1,Fin de semana 7:00-9:59,161,7.766667,13.473054


### 4.3 Registro de decisión (8 pts)

Redacta el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Conversión**: formato utilizado y verificación.
- **Zona horaria**: en qué zona están los datos y cómo lo verificaste.
- **Criterio de exclusión**: condiciones, umbrales y justificación.
- **Resultado**: viajes antes, después y excluidos.
- **Hallazgo de 4.2** y **limitaciones** del análisis.

**Respuesta 4.3:**

**Registro de decisión de limpieza — Dataset de taxis de Nueva York (marzo 2019)**

- **Conversión**: las columnas `pickup` y `dropoff` se convirtieron de texto a fecha/hora con `pd.to_datetime()`, indicando el formato explícito `'%Y-%m-%d %H:%M:%S'`, correspondiente al patrón observado en los datos (por ejemplo, `2019-03-23 20:21:09`). Se verificó que el número de valores faltantes (`NaT`) fuera el mismo antes y después de la conversión (0 en ambos casos), confirmando que todas las cadenas de texto se interpretaron correctamente.

- **Zona horaria**: los datos están registrados en **hora local de Nueva York (`America/New_York`)**, tal como indica el contexto de la actividad. Esto se verificó de forma independiente al observar que, el 10 de marzo de 2019 (día del cambio a horario de verano), **ningún viaje tiene `pickup` a las 2:00 horas** entre los 185 registrados ese día, y al localizar `pickup` en esa zona horaria con `tz_localize(nonexistent='NaT')` no se generó ningún `NaT`. Ambos resultados son consistentes con datos que ya reflejan el salto de las 02:00 a las 03:00 propio del cambio de horario, y no con datos en UTC o en otra zona sin ajustar.

- **Criterio de exclusión**: se excluyeron los viajes que cumplen **al menos una** de estas condiciones: (1) `pickup` fuera de marzo de 2019 (1 viaje, residuo del corte del periodo); (2) duración menor a 1 minuto (72 viajes, que incluyen tanto viajes de distancia cero como viajes con distancia positiva pero duración de segundos, ambos incompatibles con un desplazamiento real); (3) velocidad promedio mayor a 60 mph (11 viajes, velocidades no plausibles en tráfico urbano). Ningún viaje presentó duración negativa ni duración mayor a 3 horas.

- **Resultado**: el dataset pasó de 6,433 a 6,359 viajes, es decir, se excluyeron 74 viajes en total (1.15% del dataset). Se verificó que `taxis_limpio` no contiene viajes fuera de marzo de 2019, con duración menor a 1 minuto, ni con velocidad mayor a 60 mph.

- **Hallazgo de 4.2**: al comparar los viajes matutinos (7:00–9:59) entre días hábiles y fin de semana sobre `taxis_limpio`, los días hábiles tienen más del cuádruple de viajes (683 contra 161) y una duración mediana mayor (12.03 contra 7.77 minutos), pero una **velocidad mediana menor** (9.10 contra 13.47 mph). Esto es consistente con la congestión típica de la hora pico entre semana, que alarga los trayectos y reduce la velocidad promedio, en contraste con un fin de semana con menos tráfico.

- **Limitaciones**: el umbral de velocidad (60 mph) y el de duración mínima (1 minuto) son reglas prácticas basadas en el conocimiento general del tráfico urbano y en la inspección de los datos, no en un estándar oficial de la TLC; un análisis más riguroso podría contrastar estos umbrales con las velocidades máximas permitidas en las vías de Nueva York o con reportes similares de la propia TLC. Además, no fue posible verificar con un identificador único si los viajes de duración muy corta correspondieron a cancelaciones reales o a errores del sistema de facturación, ya que el dataset no incluye esa información.


---
## Reflexión final (10 puntos)

1. ¿Qué ventajas tuvo indicar el formato explícito en 1.1 frente a dejar que pandas lo infiera? ¿En qué situación la inferencia habría producido errores silenciosos?
2. Supón que `pickup` se hubiera registrado en UTC y `dropoff` en hora local de Nueva York. ¿Qué efecto tendría en las duraciones calculadas? ¿Cómo lo habrías detectado con las herramientas de esta actividad?
3. Si se entrenara un modelo para predecir la duración de un viaje, ¿qué problemas causarían las anomalías temporales que encontraste si no se limpiaran?

**Respuesta final:**

1. **Ventaja del formato explícito en 1.1**: indicar el formato exacto (`'%Y-%m-%d %H:%M:%S'`) hace que `pd.to_datetime()` interprete cada valor de forma determinista y más rápida, y que **falle explícitamente** (con un error) si algún valor no calza con ese patrón, en lugar de intentar adivinarlo. Si se hubiera dejado que pandas infiriera el formato, el riesgo principal habría sido con fechas ambiguas entre **día y mes** (por ejemplo, `03/04/2019` podría interpretarse como 3 de abril o como 4 de marzo dependiendo de la convención). En un dataset con fechas como `01/02/2019` a `12/03/2019`, la inferencia automática podría cambiar de convención (día/mes vs. mes/día) sin avisar, produciendo fechas válidas pero **incorrectas**: un error silencioso que no generaría ningún `NaT` y por lo tanto pasaría desapercibido en la verificación de faltantes.

2. **Efecto de mezclar UTC y hora local**: si `pickup` estuviera en UTC y `dropoff` en hora local de Nueva York (UTC-4 o UTC-5 según la época del año), al restar `dropoff - pickup` directamente (sin convertir ambas a la misma referencia horaria) la diferencia calculada estaría desplazada por esas 4 o 5 horas, dando duraciones absurdamente largas (o incluso negativas, si el desfase hiciera que el `dropoff` pareciera anterior al `pickup`). Esto se habría detectado con las mismas herramientas usadas en la Actividad 2: al calcular `duracion_min` y revisar el `describe()`, se habrían observado una media y un máximo mucho más altos de lo razonable (cercanos a 240–300 minutos de desfase sistemático), y al contar duraciones negativas o mayores a 3 horas (2.2), esos conteos habrían sido anormalmente altos y afectando a la mayoría de las filas en vez de a un puñado de casos aislados, lo que habría sido una señal clara de un problema de zona horaria y no de errores puntuales de captura.

3. **Problemas para un modelo predictivo si no se limpian las anomalías**: entrenar un modelo de duración de viaje con los datos originales generaría varios problemas. Los viajes de duración cero o menor a 1 minuto (con tarifas normales o incluso altas) introducirían **ruido y relaciones falsas** entre las variables de entrada (distancia, tarifa) y la duración, dificultando que el modelo aprenda el patrón real. Los viajes con velocidad extrema (errores de distancia o de tiempo) actuarían como **valores atípicos influyentes** que distorsionan los coeficientes o las particiones del modelo, especialmente en algoritmos sensibles a outliers (como la regresión lineal). En conjunto, esto llevaría a un modelo con **mayor error de predicción** y con métricas de evaluación poco representativas del desempeño real sobre viajes típicos, además de que en producción el modelo podría generar predicciones erráticas si estos patrones anómalos se repiten en datos nuevos sin ser detectados.


---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.